Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLSANSReduction.py

```diff
@@ -0,0 +1,331 @@
+        elif normalise_by == 'Timer':
+            if mtd[ws].getRun().hasProperty('timer'):
+                duration = mtd[ws].getRun().getLogData('timer').value
+                if duration != 0.:
+                    Scale(InputWorkspace=ws, Factor=1./duration, OutputWorkspace=ws)
+                else:
+                    raise RuntimeError('Unable to normalise to time; duration found is 0 seconds.')
+            else:
+                raise RuntimeError('Normalise to timer requested, but timer information is not available.')
+
+    def _process_beam(self, ws):
+        """
+            Calculates the beam center's x,y coordinates, and the beam flux
+            @param ws : the input [empty beam] workspace
+        """
+
+        centers = ws + '_centers'
+        method = self.getProperty('DirectBeam').value
+        radius = self.getProperty('BeamRadius').value
+        FindCenterOfMassPosition(InputWorkspace=ws, DirectBeam=method, BeamRadius=radius, Output=centers)
+        beam_x = mtd[centers].cell(0,1)
+        beam_y = mtd[centers].cell(1,1)
+        AddSampleLog(Workspace=ws, LogName='BeamCenterX', LogText=str(beam_x), LogType='Number')
+        AddSampleLog(Workspace=ws, LogName='BeamCenterY', LogText=str(beam_y), LogType='Number')
+        DeleteWorkspace(centers)
+        MoveInstrumentComponent(Workspace=ws, X=-beam_x, Y=-beam_y, ComponentName='detector')
+        integral = self._integrate_in_radius(ws, radius)
+        run = mtd[ws].getRun()
+        att_coeff = 1.
+        if run.hasProperty('attenuator.attenuation_coefficient'):
+            att_coeff = run.getLogData('attenuator.attenuation_coefficient').value
+        elif run.hasProperty('attenuator.attenuation_value'):
+            att_coeff = run.getLogData('attenuator.attenuation_value').value
+        self.log().notice('Found attenuator coefficient/value: {0}'.format(att_coeff))
+        Scale(InputWorkspace=integral, Factor=att_coeff, OutputWorkspace=integral)
+        AddSampleLog(Workspace=ws, LogName='BeamFluxValue', LogText=str(mtd[integral].readY(0)[0]), LogType='Number')
+        AddSampleLog(Workspace=ws, LogName='BeamFluxError', LogText=str(mtd[integral].readE(0)[0]), LogType='Number')
+        DeleteWorkspace(integral)
+
+    def _check_distances_match(self, ws1, ws2):
+        tolerance = 0.01 #m
+        l2_1 = mtd[ws1].getRun().getLogData('L2').value
+        l2_2 = mtd[ws2].getRun().getLogData('L2').value
+        return fabs(l2_1 - l2_2) < tolerance
+
+    def PyExec(self): # noqa: C901
+
+        process = self.getPropertyValue('ProcessAs')
+        ws = '__' + self.getPropertyValue('OutputWorkspace')
+        LoadAndMerge(Filename=self.getPropertyValue('Run').replace(',','+'), LoaderName='LoadILLSANS', OutputWorkspace=ws)
+        self._normalise(ws)
+        ExtractMonitors(InputWorkspace=ws, DetectorWorkspace=ws)
+        if process != 'Absorber':
+            absorber_ws = self.getPropertyValue('AbsorberInputWorkspace')
```

Comment:
Using `getPropertyValue()` will fail if the workspace is not in the ADS since you'll get an empty string. The same problem is lurking in a other few places in this file, as well.